src/Network/Transport/Chan.hs

```diff
@@ -29,72 +67,212 @@ data TransportState = State { _channels         :: Map EndPointAddress (Chan Eve
-apiNewEndPoint :: MVar TransportState -> IO (Either (TransportError NewEndPointErrorCode) EndPoint)
-apiNewEndPoint state = do
-  chan <- newChan
-  addr <- modifyMVar state $ \st -> do
-    let addr = EndPointAddress . BSC.pack . show . Map.size $ st ^. channels
-    return ((channelAt addr ^= chan) . (nextConnectionIdAt addr ^= 1) $ st, addr)
-  return . Right $ EndPoint { receive       = readChan chan
-                            , address       = addr
-                            , connect       = apiConnect addr state
-                            , closeEndPoint = throwIO (userError "closeEndPoint not implemented")
-                            , newMulticastGroup     = apiNewMulticastGroup state addr
-                            , resolveMulticastGroup = apiResolveMulticastGroup state addr
-                            }
+  state <- newTVarIO $ TransportValid $ ValidTransportState
+    { _localEndPoints = Map.empty
+    , _nextLocalEndPointId = 0
+    }
+  return Transport
+    { newEndPoint    = apiNewEndPoint state
+    , closeTransport = do
+        -- transactions are splitted into smaller ones intentionally
+        old <- atomically $ swapTVar state TransportClosed
+        case old of
+          TransportClosed -> return ()
+          TransportValid tvst -> do
+            forM_ (tvst ^. localEndPoints) $ \l -> do
+              cons <- atomically $ whenValidLocalEndPointState l $ \lvst -> do
+                writeTChan (localEndPointChannel l) EndPointClosed
+                writeTVar  (localEndPointState l) LocalEndPointClosed
+                return (lvst ^. connections)
+              forM_ (cons :: Map (EndPointAddress, ConnectionId) LocalConnection) $ \con ->
+                atomically $ writeTVar (localConnectionState con) LocalConnectionClosed
+    }
+
+-- | Create a new end point.
+apiNewEndPoint :: TVar TransportState -> IO (Either (TransportError NewEndPointErrorCode) EndPoint)
+apiNewEndPoint state = handle (return . Left) $ atomically $ do
+  chan <- newTChan
+  (lep,addr) <- withValidTransportState state NewEndPointFailed $ \vst -> do
+    lepState <- newTVar $ LocalEndPointValid $ ValidLocalEndPointState
+      { _nextConnectionId = 1
+      , _connections = Map.empty
+      , _multigroups = Map.empty
+      }
+    let r = nextLocalEndPointId ^: (+ 1) $ vst
+        addr = EndPointAddress . BSC.pack . show $ r ^. nextLocalEndPointId
+        lep = LocalEndPoint
+          { localEndPointAddress = addr
+          , localEndPointChannel = chan
+          , localEndPointState = lepState
+          }
+    writeTVar state (TransportValid $! localEndPointAt addr ^= Just lep $ r)
+    return (lep, addr)
+  return $ Right $ EndPoint
+    { receive       = atomically $ do
+        result <- tryReadTChan chan
+        case result of
+          Nothing -> do st <- readTVar (localEndPointState lep)
+                        case st of
+                          LocalEndPointClosed -> throwSTM (userError "Channel is closed.")
```

Comment:
As far as I remember n-t do not specify what kind of exception should be thrown if receive is called on the closed transport. `ConnectionClosed` is about lightweight connection. N-t-tcp seems to use userError "Endpoint closed".